### PR TITLE
Make TimeGraph optional in CaptureWindow.

### PR DESCRIPTION
**Pull request by florian-kuebler**
TimeGraph contains a lot of information/logic to draw capture
related information in the CaptureWindow. However, if we don't have
a capture yet, it will still be queried, e.g. for the capture duration.

This let to a couple of rendering bugs in the past. With this change
the time_graph of CaptureWindow is a unique_ptr (optional does not work
as TimeGraph is not easily movable) and is nullptr before a capture
was taken/loaded.

This also has the nice effect that the capture window is really
clean before taking a capture (no scrollbar, no time track, no
vertical green line).

Test: Try different actions (like clicks) on the empty capture window.
Load a capture and take a caputre.

### CODE REVIEW

**Review by reichlfl — APPROVED** [review]
Awesome, thanks for cleaning this up! It's hard to judge if anything was missed in CaptureWindow.cpp, looks good as far as I can tell. Did you make sure to test with both captures from file and connected to an instance?

I have a few minor comments, but I guess the only thing that should be addressed is the TODO, the rest LGTM!

File: src/OrbitGl/CaptureWindow.h [hunk 1]
```diff
@@ -72,7 +72,14 @@ class CaptureWindow : public GlCanvas {
...
   void ToggleDrawHelp();
   void set_draw_help(bool draw_help);
-  [[nodiscard]] TimeGraph* GetTimeGraph() { return &time_graph_; }
+  [[nodiscard]] TimeGraph* GetTimeGraph() {
+    if (time_graph_ == nullptr) return nullptr;
+    return time_graph_.get();
+  }
+  void CreateTimeGraph(CaptureData* capture_data) {
+    time_graph_ = std::make_unique<TimeGraph>(app_, &text_renderer_, this, capture_data);
+  }
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/OrbitGl/CaptureWindow.h]: Done.

File: src/OrbitGl/CaptureWindow.h [hunk 2]
```diff
@@ -72,7 +72,14 @@ class CaptureWindow : public GlCanvas {
   virtual void ToggleRecording();
   void ToggleDrawHelp();
   void set_draw_help(bool draw_help);
-  [[nodiscard]] TimeGraph* GetTimeGraph() { return &time_graph_; }
+  [[nodiscard]] TimeGraph* GetTimeGraph() {
+    if (time_graph_ == nullptr) return nullptr;
+    return time_graph_.get();
+  }
+  void CreateTimeGraph(CaptureData* capture_data) {
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/OrbitGl/CaptureWindow.h]: I tried this, should work, but is a rather large change. Created b/180313647 for this.

File: src/OrbitGl/TimeGraph.cpp [hunk 3]
```diff
@@ -39,9 +39,15 @@ using orbit_client_protos::CallstackEvent;
...
-    : accessibility_(this), batcher_(BatcherId::kTimeGraph), app_{app} {
-  track_manager_ = std::make_unique<TrackManager>(this, app);
+TimeGraph::TimeGraph(OrbitApp* app, TextRenderer* text_renderer, GlCanvas* canvas,
+                     CaptureData* capture_data)
+    : text_renderer_{text_renderer},
+      accessibility_(this),
+      batcher_(BatcherId::kTimeGraph),
+      capture_data_{capture_data},
+      app_{app} {
+  SetCanvas(canvas);
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/OrbitGl/TimeGraph.cpp]: Removed it. Thanks!

File: src/OrbitGl/CaptureWindow.cpp [hunk 4]
```diff
@@ -763,10 +801,10 @@ void CaptureWindow::RenderHelpUi() {
 
   Vec2 text_bounding_box_pos;
   Vec2 text_bounding_box_size;
-  text_renderer_.AddText(GetHelpText(), world_x, world_y, GlCanvas::kZValueTextUi,
-                         Color(255, 255, 255, 255), time_graph_.GetLayout().GetFontSize(),
-                         -1.f /*max_size*/, false /*right_justified*/, &text_bounding_box_pos,
-                         &text_bounding_box_size);
+  text_renderer_.AddText(
+      GetHelpText(), world_x, world_y, GlCanvas::kZValueTextUi, Color(255, 255, 255, 255),
+      14 /*TODO time_graph_->GetLayout().GetFontSize()*/, -1.f /*max_size*/,
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/OrbitGl/CaptureWindow.cpp]: Good catch, actually wanted to have another look before creating the PR, but I don't see a good way for now.
Created b/180312795 for this and referenced it in the code.

**Review by karupayun — APPROVED** [review]
LGTM Florian! :) Thanks for this!

All seems perfect. I'm only not 100% sure about making it std::optional instead, but as you already mentioned it in the PR message, I will trust in your criteria.

**Comment by florian-kuebler** [comment]
> LGTM Florian! :) Thanks for this!
> 
> All seems perfect. I'm only not 100% sure about making it std::optional instead, but as you already mentioned it in the PR message, I will trust in your criteria.

I wished it to be optional, but it is just not worth it. We would need to make a lot of classes that are explicitly neither movable nor copiable at least movable and deal with the mutex (which is also not movable!). It is probably doable but requires a lot of effort.

**Pull request merged**